docs,okeps: Localnet API proposal (#5087)
ovn-kubernetes/ovn-kubernetes: docs/okeps/okep-5085-localnet-api.md

> # OKEP-5085: Localnet API

Author: The localnet naming is well established in the project and D/S systems, and users I know of, I am afraid changing the name would cause confusion.

@maiqueb @phoracek WDYT?

> # OKEP-5085: Localnet API

Reviewer: I share Or's opinion. I would prefer to stick to localnet. This term was used by our existing users through OVN Kubernetes secondary networks since November 2023 and is widely adopted. Also, most of the users I interact with are coming from VMware, so switch to OpenStack vocabulary would not help them. If we switched to "provider" now, "localnet" would not go away.

> # OKEP-5085: Localnet API

Reviewer 2: ok, I figured...but it was worth a shot :)

`spec.desiredState.ovn.bridge-mappings` name:
```yaml
spec:
  desiredState:
      ovn:
        bridge-mappings:
        - localnet: physnet  <---- has to match the NAD config.spec.name. OR 
                                   the NAD config.spec.physicalNetworkName
          bridge: br-ex <--------- OVS switch
```
* To overcome this, and avoid exposing the net-conf-network-name in the CUDN CRD spec a new field should be introduced.
  The new field should allow users pointing to the bridge-mapping network-name they defined in the node.
  The field should be translated to the CNI physicalNetworkName field.

##### MTU
By default OVN-K sets the pod interface with MTU 1400, unless otherwise specified.

For localnet topology this is not optimal because the host physical network-interface usually has default MTU 1500.
In case no MTU is specified the pod interface connected to the host localnet is configured with MTU 1400 resulting
in bad performance and bad UX requiring troubleshooting the MTU alignment across the stack.

As long OVN-K does not set default MTU (1500) for localnet topology network,

> the controller should specify MTU to be 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

Author: As of today the default in ovn-k is defined by the platform and not by the API, I followed the same practice.
From what I understand, having controller set the defaults enable better backward compatibility in case default should change, as the controller can have additional logic to handle the new default.
Please see https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#defaulting

> the controller should specify MTU to be 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

Reviewer 2: Hmm I actually agree with this approach, since it might be possible for OVNK to detect the physical underlay MTU of the bridge and then automatically default the MTU to that. However, I think we should be clear and explicit about this in the enhancement here. So rewording what you have to something like:

By default OVN-K sets the MTU of the UDN to be 100 bytes less than the physical MTU of the underlay network. For the localnet topology this is not optimal because localnet does not use a Geneve overlay and is directly connected to the underlay. This results in a loss in throughput and potential MTU mismatch issues.

The MTU value may be set by the user, and  if not set then OVN-Kubernetes will determine the default value to use.

> the controller should specify MTU to be 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

Author: Done, PTAL

#### Workflow Description

> The CUDN CRD controller should be changes accordingly to support localnet topology.

Author: Fixed

It should validate localnet topology configuration and generate corresponding NADs for localnet as other topologies (Layer2 & Layer3)
in the selected namespaces.

As of today the CUDN CRD controller does not allow mutating `spec.network`.

OVN-Kubernetes support NAD spec mutations on day 2.
For the new network configuration (defined in the spec) to take place, the user must restart the workloads.

In order to comply with NAD spec mutation support and provide better UX,
the CUDN CRD controller should allow `spec.network` mutations for localnet topologies.

> Enable changing at least the following fields: `MTU`, `VLAN`, `excludeSubnets` and `physicalNetworkName`.

Author: Sorry for confusion.
I tried to express an exception for localnet topology, w/o getting into much of details.
Should I remove the non-goal or rephrase it?

> Enable changing at least the following fields: `MTU`, `VLAN`, `excludeSubnets` and `physicalNetworkName`.

Reviewer 3: Please remove the non-goal. 

I've expressed this concern as well in the past.

> Enable changing at least the following fields: `MTU`, `VLAN`, `excludeSubnets` and `physicalNetworkName`.

Author: Done

In any order:
- The user configures localnet bridge mapping on the nodes, e.g.: using NNCP.
- Create CUDN, specifying the bridge-mapping network name in the spec.

#### Generating the NAD
##### OVS bridge-mapping’s network-name
Introduce attribute to point ovs-bridge bridge-mapping network-name.
This attribute name should be translated to the CNI “physicalNetworkName” attribute.

Proposal the CUDN spec field name:  “physicalNetworkName”.

##### MTU
Should be translated to the CNI “mtu” attribute.

> If not specified the controller should specify default MTU (1500) in the NAD spec.config.

Reviewer 2: may also need some cleanup with the rewording of MTU

Author: Done

| MTU                 | The maximum transmission unit (MTU).                                                                                                                                                                                                                                                                  | Yes      |
| VLAN                | The network VLAN ID.                                                                                                                                                                                                                                                                                  | Yes      |
| Subnets             | List of CIDRs used for the pod network across the cluster.Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed. The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `ipam.mode` is `Disabled`. | Yes      |

> | ExcludeSubnets      | List of CIDRs removed from the specified CIDRs in `subnets`.The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `subnets` is unset or `ipam.mode` is `Disabled`.                                                                             | Yes      |

Author: Please note that we didn't add `excludeSubnets` to l2 & l3 typologies because there was no actually use case for it. At least this is what I know, I might miss something.
In general, since l2 & l3 secondary networks does not connect to the external network as with localnet, its not likely to require reserved IPs ,right?

Do you think we  have `excludeSubnets` in l2 & l3 outside locanet API development scope?
Maybe adding it upon user request later on?

> | ExcludeSubnets      | List of CIDRs removed from the specified CIDRs in `subnets`.The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `subnets` is unset or `ipam.mode` is `Disabled`.                                                                             | Yes      |

Reviewer 2: With BGP support (and EVPN coming) you can extend a layer 2 or layer 3 UDN network into the external network. So we will need exclude subnets there as well. @jcaamano just a heads up. We can add it to those topologies later, but the behavior should be consistent across all 3 types. In other words, don't write specific implementations for each topology type, lets keep them common.

> | IPAM                | Contains IPAM-related configuration for the network. When ipam.lifecycle=Persistent enable workloads have persistent IP addresses. For example: Virtual Machines will have the same IP addresses along their lifecycle (stop, start migration, reboots).                                              | Yes      |

Author: Done
Added text about `IPAM` options.

#### Suggested API validations
- `Role`:
    - Required.
    - When `topology=Localnet`, the only allowed value is `Secondary`.
        - Having Role explicitly makes the API predictable and consistent with other topologies. In addition, it enables extending localnet to support future role options
- `PhysicalNetworkName`:
    -  Required.
    - Max length 253.
    - Cannot contain `,` or `:` characters.
- `MTU`:
    - Minimum 576 (minimal for IPv4). Maximum: 65536.
    - When Subnets consist of IPv6 CIDR, minimum MTU should be 1280.
- `VLAN`: <br/>
  According to [dot1q (IEEE 802.1Q)](https://ieeexplore.ieee.org/document/10004498),
  VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>
  The VLAN IDs `0`, `1` and `4095` are reserved. <br/>
  Suggested validations:
    - Minimum: 2, Maximum: 4094.
- `Subnets`:
    - Minimum items 1, Maximum items 2.
    - Items are valid CIDR (e.g.: "10.128.0.0/16")
    - When 2 items are specified they must be of different IP families.

> - `ExcludeSubnets`:

Author: > so excludedSubnets is a union of up to 25 entries, and that union is per address family?

Its 25 entries in total regardless the family.
We need to have some upper boundary for the entries count.
Previous discussion about the it took place [here](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4575#discussion_r1700343577).

> So for example, if I had subnets as 10.0.0.0/16, I could have excludedSubnets as "10.0.1.0/24,10.0.0.0/26", right?

Yes, the format in CR YAML would be `excludeSubnets: ["10.0.1.0/24", "10.0.0.0/26"]`

    - Minimum items 1, Maximum items - 25.
    - Items are valid CIDR (e.g.: "10.128.0.0/16")
    - Cannot be set when Subnet is unset or `ipam.mode=Disabled`.

>     - Ensure excluded subnet in range of at least on subnet in `spec.network.localnet.subents`

Author: Fixed

    // Each item should be in range of the specified CIDR(s) in `subnets`.
	// The maximal exceptions allowed is 25.
    // The format should match standard CIDR notation (for example, "10.128.0.0/16").
    // This field must be omitted if `subnets` is unset or `ipam.mode` is `Disabled`.
    // In a scenario `physicalNetworkName` points to OVS bridge mapping of a network who has reserved IP addresses
    // that shouldn't be assigned by OVN-Kubernetes, the specified CIDRs will not be assigned. For example:
    // Given: `subnets: "10.0.0.0/24"`, `excludeSubnets: "10.0.0.200/30", the following addresses will not be assigned to pods: `10.0.0.201`, `10.0.0.202`.
    // +optional
    // +kubebuilder:validation:MinItems=1
    // +kubebuilder:validation:MaxItems=25
    ExcludeSubnets []CIDR `json:"excludeSubnets,omitempty"`
    
    // ipam configurations for the network (optional).
    // IPAM is optional, when omitted, `subnets` should be specified.
    // When `ipam.mode` is `Disabled`, `subnets` should be omitted.
    // `ipam.mode` controls how much of the IP configuration will be managed by OVN.

>     //    When `Enabled`, OVN-Kubernetes will apply IP configuration to the SND infra and assign IPs from the selected subnet to the pods.

Author: Fixed

   - Update API reference docs.
2. Enable localnet CUDN CRs spec mutation 
   - Invert the CRD validations in a way spec mutation is allowed for localnet topology configuration only.
   - Add support for localnet topology configuration mutations in the CUDN CRD controller.
3. Introduce CEL validation rule to ensure `excludedSubnets` items are in range of specified items in `subnets`.
   - Can be done once is resolved https://github.com/kubernetes/kubernetes/issues/130441.
   - Update the Kubernetes version using in CI that includes the bugfix.
   - Add the subject validation.
4. E2e test verifying Kubevirt VMs can communicate over locanet topology network created using CUDN CR.

### Testing Details

Unit tests should cover the new localnet topology API extension, and spec mutation for localnet topology spec.

As of today localnet topology implementation is tested e2e as part of multi-homing test suite.
The proposed solution should be tested e2e verifying pods can communicate over localnet topology created using CUDN CR.
Similar e2e tests should be introduced using Kubevirt VMs.

### Documentation Details

## Risks, Known Limitations and Mitigations

CEL rule validations:

> Validating specified subnets exclusions (`spec.netwrok.localnet.excludedSubnets`) are in range of the specified topology subnets 

Author: Fixed

> (`spec.netwrok.localnet.excludedSubnets`) is currently impossible due to a bug in the CEL rule library for validation IPs and CIDRs.

Author: Fixed

`spec.desiredState.ovn.bridge-mappings` name:
```yaml
spec:
  desiredState:
      ovn:
        bridge-mappings:
        - localnet: physnet  <---- has to match the NAD config.spec.name. OR 
                                   the NAD config.spec.physicalNetworkName
          bridge: br-ex <--------- OVS switch
```
* To overcome this, and avoid exposing the net-conf-network-name in the CUDN CRD spec a new field should be introduced.
  The new field should allow users pointing to the bridge-mapping network-name they defined in the node.
  The field should be translated to the CNI physicalNetworkName field.

##### MTU
By default, OVN-K sets the MTU of the UDN to be 100 bytes less than the physical MTU of the underlay network.
For the localnet topology this is not optimal because localnet does not use a Geneve overlay and is directly 
connected to the underlay.
This results in a loss in throughput and potential MTU mismatch issues.

The MTU value may be set by the user, and if not set then OVN-Kubernetes will determine the default value to use.

> For localnet topology, the controller should specify MTU 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

Reviewer 2: actually why is this MTU section even here (under Underlying network name header)? You have an identical MTU section under

#### Generating the NAD

Seems like you should remove this section and just use the one under Generating the NAD, and remove this sentence about controller setting 1500.

> For localnet topology, the controller should specify MTU 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

Author: Sorry for the confusion, I removed this text.

Details about MTU remain under "Generating the NAD" header.

      - Due to a bug in Kubernetes CEL validation IP/CIDR operations this validation can be implemented once the following issue is resolved
        https://github.com/kubernetes/kubernetes/issues/130441 
        The CRD controller should validate excludeSubnets items are in range of specified subnets. 
        In a case of an invalid request raise an error in the status.

#### YAML examples
Assuming the node has OVS bridge-mapping defined by [Kubernetes-nmstate](https://nmstate.io/kubernetes-nmstate/) 
using the following `NodeNetworkConfigurationPolicy` (NNCP):
```yaml
...
desiredState:
    ovn:
      bridge-mappings:
      - localnet: tenantblue 
        bridge: br-ex
```
Example 1:
```yaml
---
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: test-net
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Localnet
    localnet:
      role: Secondary
      physicalNetworkName: tenantblue
      subnets: ["192.168.100.0/24", "2001:dbb::/64"]
      excludeSubnets: ["192.168.100.1/32", "2001:dbb::0/128"]
```

> The above spec will generate the following NAD, in namespace `blue`:

Author: Done

| PhysicalNetworkName | The OVN bridge mapping network name is configured on the node.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | No       |

> | MTU                 | The maximum transmission unit (MTU).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Yes      |

Reviewer 4: @ormergi If MTU is optional, is there a default value that is used?

Author: The default MTU for localnet topology should be 1500 https://github.com/ormergi/ovn-kubernetes/blob/5f86091e9bc7e787d171041c6242e4dcc7a81269/docs/okeps/okep-5085-localnet-api.md#mtu

    - Overrides the `name` attribute, defined in (1).
    - Allows multiple localnet topology NADs to refer to the same bridge-mapping (thus simplifying the admin’s life -
      fewer manifests to provision, and keep synced).
3. `subnets`
   Subnets to use for the network across the cluster.
4. `excludeSubnets`
   IP addresses range to exclude from the assignable IP address pool specified by subnets field.
5. `vlanID` - VLAN tag assigned to traffic.
6. `mtu` - maximum transmission unit for a network
7. `allowPersistentIPs`
   persist the OVN Kubernetes assigned IP addresses in a `ipamclaims.k8s.cni.cncf.io` object. This IP addresses will be
   reused by other pods if requested. Useful for [KubeVirt](https://kubevirt.io/) VMs.

#### Extend ClusterUserDefinedNetwork CRD
Given the CUDN CRD is targeted for cluster-admin users, it is a good fit for operations that require cluster-admin intervention,
such as localnet topology.

The suggested solution is to extend the CUDN CRD to enable creating localnet topology networks.

##### Underlying network name
Given the CUDN API doesn’t expose the underlying network name, represented by the NAD `spec.config.name` (net-conf-network-name),

> by design.

Author: I will rephrase this part.

Localnet topology requires the net-conf-network-name to match the OVN bridge-mapping network-name on the node.
In case [Kubernetes-nmstate](https://nmstate.io/kubernetes-nmstate/) is used, the NAD `spec.config.name` has to match the `NodeNetworkConfigurationPolicy`
`spec.desiredState.ovn.bridge-mappings` name:
```yaml
spec:
  desiredState:
      ovn:
        bridge-mappings:
        - localnet: physnet  <---- has to match the NAD config.spec.name. OR 
                                   the NAD config.spec.physicalNetworkName
          bridge: br-ex <--------- OVS switch
```
* To overcome this, and avoid exposing the net-conf-network-name in the CUDN CRD spec a new field should be introduced.
  The new field should allow users pointing to the bridge-mapping network-name they defined in the node.
  The field should be translated to the CNI physicalNetworkName field.

> to have indication about which workload require restart following spec change, thus allowing the workload to consume the latest network config

Author: This is a left over that is not longer relevant, I will remove it. Sorry for the confusion.

      excludeSubnets: ["192.168.50.0/24"]
      mtu: 9000
      vlan:
        mode: Access
        access:
          id: 200
      ipam:
        lifecycle: Persistent
```
The above spec will generate the following NAD, in namespace `red`:
```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: test-net
  namespace: red
finalizers:
 - k8s.ovn.org/user-defined-network-protection
labels:
  k8s.ovn.org/user-defined-network: ""
ownerReferences:
- apiVersion: k8s.ovn.org/v1
  blockOwnerDeletion: true
  controller: true
  kind: ClusterUserDefinedNetwork
  name: test-net
spec:
    config: > '{
        "cniVersion": "1.0.0",
        "type": "ovn-k8s-cni-overlay"
        "netAttachDefName": "blue/test-net",
        "role": "secondary",
        "topology": "localnet",
        "name": "cluster.udn.test-net",
        "physicalNetworkName: "tenantblue",
        "subnets": "192.168.0.0/16,2001:dbb::/64",
        "allowPersistentIPs": true,
        "excludesubnets: "10.100.50.0/24",
        "mtu": 9000,
        "vlanID": 200
    }'
```

### Implementation Details

The CUDN `spec.network.topology` field should be extended to accept `Localnet` string.

> And the `spec.network` struct `NetworkSpec` should be to have an additional field for localnet topology configuration:

Reviewer 5: ```suggestion
And the `spec.network` struct `NetworkSpec` should have an additional field for localnet topology configuration:
```

    ...
    // Localnet is the Localnet topology configuration.
    // +optional
    Localnet *LocalnetConfig `json:"localnet,omitempty"`
}
```

The CUDN spec should have additional validation rule for `spec.network.topology` field:
```go
// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
type ClusterUserDefinedNetworkSpec struct {
    ...
    // +required
    Network NetworkSpec `json:"network"`
}
```

#### Localnet topology configuration type
Introduce new topology configuration type for localnet - `LocalnetConfig`.
The CUDN CRD `spec.network` should feature proposed localnet topology configuration.

The Layer2 and Layer3 configuration types (`Layer2Config` & `Layer3Config`) `role` field is defined with `NetworkRole` type.
The `NetworkRole` type has the following enum validation, allowing `Secondary` and `Primary` values:
```
// +kubebuilder:validation:Enum=Primary;Secondary
```
The proposed localnet config type (`LocalnetConfig`) `role` field would have to accept `Secondary` value only.
In order to avoid misleading CRD scheme, the enum validation should be moved closer to each `NetworkRole` usage:

> 1. Remove the enum validation exist on `NetworkRole` definition

Reviewer 5: If I understood correctly:
```suggestion
1. Remove the enum validation `exists` from `NetworkRole` definition
```

Author: I meant to say we need to remove the existing validation.
I will rephrase this part.

2. At the `Layer2Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
3. At the `Layer3Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
4. At the proposed `LocalnetConfig` definition, add enum validation to `role` field allowing `Secondary` only.

#### VLAN configuration type

> As of today OVN-Kubernetes CNI enable setting the access VLAN of a localnet topology the network (using NADs).

Reviewer 5: ```suggestion
As of today OVN-Kubernetes CNI allows to set the access VLAN of a localnet topology (using NADs).
```

There are two options to expose the VLAN attribute:
1. Single integer filed, living at the same level as other fields, for example:
   ```
    localnet:
       mtu: 1500
       vlan: 100
   ```
2. Discriminated union for accommodating VLAN related configurations, for example: 
   ```
    localnet:
       mtu: 1500
       vlan: 
         mode: Access
         access:
           id: 100 
   ```

> Although option (1) being pretty straight forward, in case additional VLAN configurations should be supported requiring 

Reviewer 5: ```suggestion
Although option (1) is pretty straightforward, if additional VLAN configurations need to be supported, requiring 
```

> exposing additional fields, it will enforce having VLAN related filed right next to non VLAN related ones, making the API look occurred.

Reviewer 5: ```suggestion
the exposure of additional fields, it will enforce having VLAN-related fields right next to non-VLAN-related ones, making the API look awkward.
```

In addition, in case future VLAN additions introduce mutual exclusive relation between VLAN config related fields,

> the flat structure make it harder to maintain the related API validations markers. (not centralized, lives alongside other validations make them harder to follow).

Reviewer 5: ```suggestion
the flat structure makes it harder to maintain the related API validations markers. (it is not centralized, lives alongside other validations and makes them harder to follow).
```

> Option (2) enable all related VLAN configuration lives under the same roof enable the API to evolve nicely, introducing less noise comparing to (1).

Reviewer 5: ```suggestion
Option (2) enables all related VLAN configurations to live under the same roof, allowing the API to evolve smoothly while introducing less noise compared to (1).
```

> And enable having all related API validations centralized.

Reviewer 5: To merge with the previous sentence:
```suggestion
 and enabling the centralization of all related API validations.
```

For example [kubernetes-nmstate](https://github.com/nmstate/kubernetes-nmstate?tab=readme-ov-file) (utilize https://nmstate.io/) 

> follow the same convention, that is having complex filed for VLAN configurations. 

Reviewer 5: ```suggestion
follows the same convention, that is having a complex field for VLAN configurations. 
```

    // When set, OVN-Kubernetes will apply VLAN configuration to the SDN infra and to the connected pods.
    // +optional
    VLAN VLANConfig `json:"vlan,omitempty"`
}
```

#### Implementation phases
1. Add support for localnet topology on top CUDN CRD
   - Extend the CUDN CRD.
   - Add support for localnet topology in the CUDN CRD controller.
   - Adjust CI multi-homing jobs to enable testing localnet CUDN CRs.
   - Update API reference docs.
2. Introduce CEL validation rule to ensure `excludedSubnets` items are in range of specified items in `subnets`.
   - Can be done once is resolved https://github.com/kubernetes/kubernetes/issues/130441.
   - Update the Kubernetes version using in CI that includes the bugfix.
   - Add the subject validation.

### Testing Details
The controller business logic is agnostic to topology types, the proposed solution can be covered fully by unit test.
In addition, the CRD controller business logic, and localnet topology functionality (using NADs) are tested e2e thoroughly. 
E2e test for localnet topology CUDN CR is optional in this particular case.

### Documentation Details

## Risks, Known Limitations and Mitigations

### CEL rule validations

> Validating specified subnets exclusions (`spec.network.localnet.excludedSubnets`) are in range of the specified topology subnets 

Reviewer 5: ```suggestion
Validating that the specified subnets exclusions (`spec.network.localnet.excludedSubnets`) are in the range of the specified topology subnets 
```

(`spec.network.localnet.excludedSubnets`) is currently impossible due to a bug in the CEL rule library for validation IPs and CIDRs.
Such invalid CUDN CRs request will not be blocked by the cluster API.
Once the bug is resolved the validation can be added.
See the following issue for more details https://github.com/kubernetes/kubernetes/issues/130441.

To mitigate this, the mentioned validation should be done by the CUDN CRD controller:

> In a scenario CUDN CR who has at least one exclude-subnet is not in range of the topology subnet,

Reviewer 5: ```suggestion
In a scenario where a CUDN CR has at least one exclude-subnet that is not within the range of the topology subnet,
```

> the controller will not create the corresponding NAD, and report an error in the status.

Reviewer 5: ```suggestion
the controller will not create the corresponding NAD and will report an error in the status.
```

# OKEP-5085: Localnet API

## Problem Statement

As of today one can create a user-defined network over localnet topology using NetworkAttachmentDefinition (NAD).
Using NAD for localnet has some pitfalls due to the fact it is not managed and not validated on creation.
Misconfigurations are detected too late causing bad UX and frustration for users.

Configuring localnet topology networks require changes to cluster nodes network stack and involve some risk and
knowledge that require cluster-admin intervention.
Such as configuring the OVS switch to which the localnet network connects to, align MTU across the stack and 
configure the right VLANs that fit the provider network.

## Goals

- Enable creating user-defined-networks over localnet topology using OVN-K CUDN CRD.

> - Streamline localnet UX: detect misconfigurations early, communicate issues via informative status conditions and or events.

Author: I rephrase the text as suggested, PTAL.

- Allow localnet topology spec changes on day2, on top the CUDN CRD.

> - Indicate whether workloads require restart following localnet topology spec change on day2.

Reviewer 3: I agree: we want to decouple finding the information from acting upon the information, *and* we don't want to specify restart - we want to identify workloads attached to UDNs with an out-dated config. 

Good catch.

> - Indicate whether workloads require restart following localnet topology spec change on day2.

Author: I rephrase the text as suggested, PTAL.

- As a user I want to be able to connect my workloads (pod/VMs) to the localnet the admin created in my namespace.
- As a user I want my workloads to be able to communicate with each other over the localnet network.
- As a user I want my connected VMs to the localnet network to be able to migrate from one node to another, having its localnet network interface IP address unchanged.
- As an admin I want to be able to perform day2 changes to the localnet topology spec following env changes.
  - Change VLAN, e.g: following the being used has been decommissioned.
  - Changing MTU, e.g: following HW has been changed requiring MTU re-alignment.
  - Change the localnet topology bridge-mapping, e.g.: following a requirement to connect to different interfaces on the node.
  - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 
    because the external network introduced a new service requiring an additional reserved IP address.
- As an admin / support engineer I want to be able to list workloads that require restart following network configuration change,

>   allowing the workloads to latest consume network configuration.

Author: I changed text to accout for the workload-manager entity and removed text about restarting workloads. PTAL

#### Suggested API validations
- `Role`:
    - Required.
    - When `topology=Localnet`, the only allowed value is `Secondary`.
        - Having Role explicitly makes the API predictable and consistent with other topologies. In addition, it enables extending localnet to support future role options
- `PhysicalNetworkName`:
    -  Required.
    - Max length 253.
    - Cannot contain `,` or `:` characters.
- `MTU`:
    - Minimum 576 (minimal for IPv4). Maximum: 65536.
    - When Subnets consist of IPv6 CIDR, minimum MTU should be 1280.
- `VLAN`: <br/>
  According to [dot1q (IEEE 802.1Q)](https://ieeexplore.ieee.org/document/10004498),
  VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>
  The VLAN IDs `0`, `1` and `4095` are reserved. <br/>
  Suggested validations:

>     - Minimum: 2, Maximum: 4094.

Reviewer 3: @RamLavi FYI

>     - Minimum: 2, Maximum: 4094.

Author: Done 
PTAL at the current state and options

  The field should be translated to the CNI physicalNetworkName field.

to have indication about which workload require restart following spec change, thus allowing the workload to consume the latest network config

#### Indicate workloads network-configuration is out of sync
Consider the following scenario:
Given localnet CUDN CR, and multiple pods connected to the network.
The CUDN CR localnet spec is changed, the controller updated the corresponding NAD spec accordingly.
Now the connected pods network-configuration is out-dated.
Re-creating the pods will allow them to consume the latest network-configuration.

Users should be able to check which connected pod is configured with outdated network-configuration,
allow them to recreate the workloads and consume the latest network-configuration.

Provide indication for users to check which workload configured with outdated network-configuration, 
following the network configuration has been changed (i.e.: CUDN CR network spec change).
The indication should enable users check which workloads are configured with an outdated network-configuration, 
save wasting time on troubleshooting, and act on it.

> The proposed solution is to set an annotation on pods who are configured with outdated network-configuration: 

Reviewer 3: > I think we should also add a note here that not every CUDN update actually requires an action from the pod's owner (with a reference to the update consequences for different fields), but we choose to put this annotation on every update to simplify our lives. 

I agree - keeping it simple surely brings benefits.

> I would also add that being more precise in the future (e.g. not setting the annotation when VLAN is updated) can be an option

This would be nice, for sure. I agree it would belong on a follow-up.

> The proposed solution is to set an annotation on pods who are configured with outdated network-configuration: 

Author: Note has been added.

@EdDev please see below "Riskes and Mitigation" for implications of changing each field. The gist of it is this: 
`VLAN`, `PhysicalNetworkName` should not introduce any disruption (i.e: disconnections).
Changing MTU may introduce network disruption (latency, TCP conn. break).
`ExcludeSubents` may introduce issues caused by IP conflict in a scenario external "service" conflict with pod IP address, because the pod has not been restarted yet (and assigned with IP address outside the exclusion pool).

  k8s.ovn.org/user-defined-network: ""
ownerReferences:
- apiVersion: k8s.ovn.org/v1
  blockOwnerDeletion: true
  controller: true
  kind: ClusterUserDefinedNetwork
  name: test-net
  uid: 293098c2-0b7e-4216-a3c6-7f8362c7aa61
spec:
    config: > '{
        "cniVersion": "1.0.0",
        "type": "ovn-k8s-cni-overlay"
        "netAttachDefName": "blue/test-net",
        "role": "secondary",
        "topology": "localnet",
        "name": "cluster.udn.test-net",
        "physicalNetworkName: "tenantblue",
        "mtu": 1500,
        "subnets": "192.168.100.0/24,2001:dbb::/64",
        "excludeSubnets": "192.168.100.1/32,2001:dbb::0/128"
    }'
```
Example 2 (custom MTU, VLAN and sticky IPs):
```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: test-net
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Localnet
    localnet:
      role: Secondary
      physicalNetworkName: tenantblue
      subnets: ["192.168.0.0/16", "2001:dbb::/64"]
      excludeSubnets: ["192.168.50.0/24"]
      mtu: 9000
      vlan:
        mode: Access
        access:

>           id: 200

Reviewer 6: This is the OVS/bridge API from nmstate:
```yaml
          vlan:
            mode: trunk
            trunk-tags:
            - id: 101
            - id-range:
                min: 500
                max: 599
            tag: 100
            enable-native: true
```
It does not make sense to have `access` as a sub-filed, because it is ok to use the `id` (or `tag`) even if the mode is a trunk (in case native is enabled).

I also do not understand why we need to invent yet another API naming. NNCP & nmstate should be followed if it makes sense. And it does make sense here (without defineing the parts that are not needed for now).

>           id: 200

Author: In case overriding the default VLAN in trunk mode (aka native vlan), the `trunk` settings can have `nativeID`, making the API more explicit and predictable.
Reusing the same filed for access mode VID and the native VID for trunk mode, is tempting because its the convention we know of.
IMO its confusing, in the above nmstate API example, looking at `tag` field  makes me wonder what the does `tag` really mean and why I need to set another boolean for it.

FWIW I tried to keep naming and structure close to nmsate API as possible.

>           id: 200

Reviewer 3: > This is the OVS/bridge API from nmstate:
> 
> ```yaml
>           vlan:
>             mode: trunk
>             trunk-tags:
>             - id: 101
>             - id-range:
>                 min: 500
>                 max: 599
>             tag: 100
>             enable-native: true
> ```
> 
> It does not make sense to have `access` as a sub-filed, because it is ok to use the `id` (or `tag`) even if the mode is a trunk (in case native is enabled).
> 
> I also do not understand why we need to invent yet another API naming. NNCP & nmstate should be followed if it makes sense. And it does make sense here (without defineing the parts that are not needed for now).

Why would we follow nmstate when we can follow bridge CNI ? 
https://www.cni.dev/plugins/current/main/bridge/#network-configuration-reference

> Well, consistency usually should win when there is no strong enough reason to break it.
The API pattern of "discriminated-unions" is overused IMO and violates other concepts like avoiding duplication.

You may not like them, but, to me, it seems they're encouraged in the openshift API guidelines: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

>           id: 200

Reviewer 6: > Why would we follow nmstate when we can follow bridge CNI ?
> https://www.cni.dev/plugins/current/main/bridge/#network-configuration-reference

Because in this case, we know in advance what are the options and on the CNI it was added as an aftermath. Same as with the IPv6 API in Kuberenetes.

> You may not like them, but, to me, it seems they're encouraged in the openshift API guidelines: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

My interpretation for the original intention is the recommendation to avoid flat fields which is hard to understand their relation. This is not that case, as one of the fields is common and has the same semantics for both valid cases.
But I do not really know the original intention, I just hope it was not a bind recommendation to always use one pattern over the other.

  - Pods should be able to communicate with each other.
  - If the external network interface MTU is higher than old MTU settings, there might be network disruption causing packet fragmentation.

- `ExcludeSubnets`
  ExcludeSubnet can be configured when `subnets` is configured, in this case OVN-Kubernetes will assigned pod IPs to pods from the specified
  value in `subnets`.
  When `ExcludeSubnet` is set, OVN-Kubernetes exclude the specified IP address from the pool and wont assign IPs in this range to pods.

  Adding or removing an exclude-subnet should not introduce network disruption whatsoever, as this setting effects OVN-Kubernetes IPAM service
  and not the pods network-stack directly.

  In a scenario the external network introduce new "service" whose IP address now considered reserved, e.g.: 10.0.0.10.
  In case there the address 10.0.0.10 already assigned a pod, both the "service" and the pod may suffer from network disruption  
  due to IP address conflict.
  Once the pod with the assigned IP is re-created, it will get different, non excluded, IP address from the pool, and the IP conflict

>   should be resolved.

Author: > I think it's OK, but we might want to consider adding a separate annotation / label to pods that we know are impacted by this. i.e. launcher pod with a VM has a duplicate address w/ a new excluded IP.

This is sounds like new feature, as off today such check, where OVN-K warns some excludeSubnet is already assigned to pods, is not implemented, right?

I hope we can keep this outside this design scope, its already creeping.
I added text about mitigation for this could be providing indication about this state, FWIW.

> Or, we may even want to prevent it in the API. Nor sure that's smart though.

I dont think its possible to enforce it at the API level using CEL validations.
Maybe using a webhook that monitors pods and NADs, but it comes with a cost and may introduce races, sounds like it deserve its own design.

>   should be resolved.

Reviewer 3: Sure, I agree. Let's limit the scope. While these could be helpful, we can also design later on.

In a scenario some workloads did not consume the latest network-config yet, while other workloads configured with latest settings,

> all workloads should still be able to communicate with each other with no issue.

Author: From what I tested on Kind cluster, pods can communicate with each other (tcp, ssh) over when one pod has lower/higher MTU than the other.
Whether they run on the same node or on different nodes. 

Although, in production where there are actual switches routers, disconnections may occur if MTU on one ends is diffrent then the other.
It may cause the switchs to fragment packers and introduce network disruptions  (e.g.: high latency, breaking TLS connections, TCP connections).

Good point, I will remove this sentence.

#### Indicate workloads network-configuration is out of sync
Consider the following scenario:
Given localnet CUDN CR, and multiple pods connected to the network.
The CUDN CR localnet spec is changed, the controller updated the corresponding NAD spec accordingly.
Now the connected pods network-configuration is out-dated.
Re-creating the pods will allow them to consume the latest network-configuration.

Users should be able to check which connected pod is configured with outdated network-configuration,
allow them to recreate the workloads and consume the latest network-configuration.

Provide indication for users to check which workload configured with outdated network-configuration, 
following the network configuration has been changed (i.e.: CUDN CR network spec change).
The indication should enable users check which workloads are configured with an outdated network-configuration, 
save wasting time on troubleshooting, and act on it.

The proposed solution is to set an annotation on pods who are configured with outdated network-configuration: 
```yaml

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Author: Sorry I dont understand what you mean by that.
Which of the key or value of the annotation you find not generic enough?

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Author: According to the convention, if a controller manage a resource and it use annotation, the annotation should be prefix with a domain.
Unprefixed annotations should be used by end-users
Please see the paragraph about annotations https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#label-selector-and-annotation-conventions

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Author: More generic solution could be adding a condition to the pod, saying OVN-Kubernetes network config is out of sync.

I am not sure its better because the condition type should be unique to avoid conflicting with other systems placing conditions.

See Alternatives section, option 2.

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Reviewer 6: > According to the convention, if a controller manage a resource and it use annotation, the annotation should be prefix with a domain.
> Unprefixed annotations should be used by end-users

I have not tried to say anything about dropping the prefix.

I did say
> have a more generic domain

and

> Use the one that is used by Multus when annotating the pods.

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Reviewer 7: I don't think we can, because this annotation is ovn-k specific, we never discussed it with a wider multi-network community.
It is an option though e.g. to make multus set the annotation on NAD spec changes, then of course the annotation key will be different, but this way will take much more time

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Reviewer 3: multus doesn't care - at all - about changes of anything.

It just runs CNI ADD / DEL / CHECK on the workloads when triggered by CRI (CNI ADD => create pod sandbox for ADD; delete pod for DEL; periodically for CHECK).

Listening to NAD spec changes would be something done on a separate controller. 
This is a major feature that would need community buy-in.

It might be reasonable to negotiate a new k8snetworkplumbingwg annotation, which OVN-K would set, and KubeVirt would consume.

> `k8s.ovn.io/network-out-of-sync='["tenant1/red", "tenant1/blue",...]`

Author: I think a middle ground solution can be using the proposed  annotation until generic one is available under k8snetplub org.

$ kubectl get pod -o jsonpath='{range .items[?(@.metadata.annotations.k8s\.ovn\.io/network-out-of-sync)]}{.metadata.name}{"\n"}{end}'
```
List pod names namespace "tenant1" who are connected to network "red" who's network-configuration is outdated:
```shell
$ kubectl get pod -o json | jq '.items[].metadata | select(.annotations."k8s.ovn.io/network-out-of-sync" | fromjson | .[] | select(. == "tenant1/red" )).name'
```

The proposed annotation for indicating whether workload configured with outdated network-config, servers as the baseline for 
additional developments for improving how this state is communicated to users. 
Please see the [Alternatives](#indicating-pod-configured-with-outdated-network-config) section for more details.

#### Workflow Description

The CUDN CRD controller should be changed accordingly to support localnet topology.
It should validate localnet topology configuration and generate corresponding NADs for localnet as other topologies (Layer2 & Layer3)
in the selected namespaces.

As of today the CUDN CRD controller does not allow mutating `spec.network`.

OVN-Kubernetes support NAD spec mutations on day 2.

> For the new network configuration (defined in the spec) to take place, the user must restart the workloads.

Author: The term restart is used all over the doc, meaning pod re-creation of VM restart that is followed by the VM pod re-creation. Sorry if it cause some confusion.

This line describe the current state in the project  (allowing NAD mutations) and how it is expected to be used, that is to re-create the workload after the NAD has been changed.

The proposed solution mitigates it in the sense only a small subset of configuration are editable.
In practice only MTU change require immediate action to prevent network issue, which we can also avoid in the future by making OVN-K CNI perform in-place MTU changes. 
See Alternative section, option 5, and Tim comment about it.

saving the hassle of re-creating the whole project (delete pods, delete network CR, create the network CR with new settings, create pods ...).

Naturally, changing network setting while workloads connected to it has implications, may introduce transient network disruption, 
which manifests in packet-drops, latency and disconnections.

Having said that mutating the suggested settings should not break the workloads networking, and they should be able to communicate
with each other.
After the network setting changes, restarting the workloads is needed in order for them to consume the latest network-config.
Once all connected workloads configured with the latest network-config the state should be reconciled and no network-disruption is expected.

Please refer to [Risks Known Limitations and Mitigations](#risks-known-limitations-and-mitigations) section for more details.  

When a CUDN CR localnet spec changes, and the change require updating the corresponding NAD spec.config, the controller should 
add the proposed annotation `k8s.ovn.io/network-out-of-sync` to all connected pods in the selected-namespaces.

> In case the annotation is already exists and has items (networks), the controller should append the subject network to the list. 

Author: The idea is to set the annotation after the NAD has been updated, so if new pods are created they will consume the new configuration no need to annotate them.
What left is the existing pods to annotate.
In case one of them is deleted its noop for the controller.

I will add text to elaborate on this point.

#### Generating the NAD
##### OVS bridge-mapping’s network-name
Introduce attribute to point ovs-bridge bridge-mapping network-name.
This attribute name should be translated to the CNI “physicalNetworkName” attribute.

Proposal the CUDN spec field name:  “physicalNetworkName”.

##### MTU
Should be translated to the CNI “mtu” attribute.

By default, OVN-K sets the MTU of the UDN to be 100 bytes less than the physical MTU of the underlay network.
For the localnet topology this is not optimal because localnet does not use a Geneve overlay and is directly
connected to the underlay.
This results in a loss in throughput and potential MTU mismatch issues.

> The MTU value may be set by the user, and if not set then OVN-Kubernetes will determine the default value to use.

Author: Default should be 1500 for localnet.
I removed the number following review comment, @trozet since you asked to remove it (at least from what I understood :sweat_smile: ), do you have objection for specify it?

#### Suggested API validations
- `Role`:
    - Required.
    - When `topology=Localnet`, the only allowed value is `Secondary`.
        - Having Role explicitly makes the API predictable and consistent with other topologies. In addition, it enables extending localnet to support future role options
- `PhysicalNetworkName`:
    -  Required.
    - Max length 253.
    - Cannot contain `,` or `:` characters.
- `MTU`:
    - Minimum 576 (minimal for IPv4). Maximum: 65536.
    - When Subnets consist of IPv6 CIDR, minimum MTU should be 1280.
- `VLAN`:

>   - `Mode`: Can be "Access".

Reviewer 6: Can be? You mean it is optional and defaults to something?

Author: Just bad wording, I meant to say is should validate the value is "Access".

  - `ID` (`access.id`)
    According to [dot1q (IEEE 802.1Q)](https://ieeexplore.ieee.org/document/10004498),
    VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>

>     The VLAN IDs `0`, `1` and `4095` are reserved. <br/>

Author: From what I saw in the do1q doc,  VID 1 is reserved for management of the appliance.
And from what I know its a common convention, for example CISCO switches use VID 1 as the defualt.
Thus I think we should not allow it in order to avoid miss-configurations.
We can always allow it upon a necessity in the future.
Make sense?

>     The VLAN IDs `0`, `1` and `4095` are reserved. <br/>

Reviewer 6: In case the node nic is connected using a trunk to the switch port, setting this ID to 1 assures it reaches that switch "default" VLAN as you mentioned. Otherwise, you will have to assume this implicitly.

>     The VLAN IDs `0`, `1` and `4095` are reserved. <br/>

Author: Good point, I will change text accordingly, to meet the future trunk mode requirements.

    - Items are valid CIDR (e.g.: "10.128.0.0/16")
    - Cannot be set when Subnet is unset or `ipam.mode=Disabled`.
    - Ensure excluded subnet in range of at least on subnet in `spec.network.localnet.subnets`
      - Due to a bug in Kubernetes CEL validation IP/CIDR operations this validation can be implemented once the following issue is resolved
        https://github.com/kubernetes/kubernetes/issues/130441 
        The CRD controller should validate excludeSubnets items are in range of specified subnets. 
        In a case of an invalid request raise an error in the status.

#### YAML examples
Assuming the node has OVS bridge-mapping defined by [Kubernetes-nmstate](https://nmstate.io/kubernetes-nmstate/) 
using the following `NodeNetworkConfigurationPolicy` (NNCP):
```yaml
...
desiredState:
    ovn:
      bridge-mappings:
      - localnet: tenantblue 
        bridge: br-ex
```
Example 1:
```yaml
---
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: test-net
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Localnet

>     localnet:

Reviewer 6: If I understand correctly, this defines a new subtree which uses fields that already appear in other topologies?
I guess that is what "discriminated-unions" means. I do not like these, IMO they hint of a lack of proper modeling.
But I guess if this is already the current pattern used, nothing we can do about it.

#### Localnet topology configuration type
Introduce new topology configuration type for localnet - `LocalnetConfig`.
The CUDN CRD `spec.network` should feature proposed localnet topology configuration.

The Layer2 and Layer3 configuration types (`Layer2Config` & `Layer3Config`) `role` field is defined with `NetworkRole` type.
The `NetworkRole` type has the following enum validation, allowing `Secondary` and `Primary` values:
```
// +kubebuilder:validation:Enum=Primary;Secondary
```
The proposed localnet config type (`LocalnetConfig`) `role` field would have to accept `Secondary` value only.
In order to avoid misleading CRD scheme, the enum validation should be moved closer to each `NetworkRole` usage:
1. Remove the enum validation exist on `NetworkRole` definition
2. At the `Layer2Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
3. At the `Layer3Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
4. At the proposed `LocalnetConfig` definition, add enum validation to `role` field allowing `Secondary` only.

#### VLAN configurations

> As of today localnet topology allows setting VLAN for the network.

Reviewer 2: you mean in the cni spec of the NAD definition right? Can you clarify that here please so that someone doesn't read this as the API you are suggesting for CUDN?

  In case there the address 10.0.0.10 already assigned a pod, both the "service" and the pod may suffer from network disruption  
  due to IP address conflict.
  Once the pod with the assigned IP is re-created, it will get different, non excluded, IP address from the pool, and the IP conflict
  should be resolved.

  Mitigation for this can be providing indication about this state - excludeSubnet cannot be set because IP in its range already
  been assigned to pods.

Once all the connected workloads consume the new network-config, no network disruption should occur as all connected workloads
network settings are aligned, using the latest network-config.

## OVN Kubernetes Version Skew

## Alternatives

### Indicating pod configured with outdated network-config
Following discussions about the proposed solution for indicating pod configured with outdated network-config,
some alternatives and improvements were suggested.
The proposed solution serves as a baseline for most suggested alternatives.
Most alternative solutions are additions serving as an additional matter for providing similar indication, as the proposed one.

> #### 1. Emit events on pods 

Reviewer 2: IMO (and i also think @jcaamano  agrees) this is a requirement and not optional for this API change. The event is a much more kubernetes way to know to see that something is wrong with a pod when a user does kubectl describe.

Reviewer 7: do you mean together with the annotation or instead of the annotation?

Author: Moved emitting event as part of the proposed solution.

Similar to the idea of the proposed solution, but with emitting events on the connected pods.
Following CUDN CR localnet spec change, the CUDN controller should emit event on connected pods.

The event should indicate pod network configuration is outdated, and suggest restarting to consume the latest network config.

> Comparing to annotation, events are harder to manage and get garbage collected.

Reviewer 2: This doesn't really apply because to remedy the problem you have to delete the pod anyway, right?

Author: I meant to express that if a pod will not re-create in order to consume the new configuration, the event may dispose and no indication is left on the pod. Unlike annotation that are just part of the metadata object.
Make sense?

In addition, filtering events to realize which pod run outdated network-config is not intuitive 
comparing to filtering by an annotation.

Emitting events is a complementary matter for indicating the subject state, alongside the proposed solution.
It can be done as a follow-up upon necessity and feedback from users.

#### 2. Add condition to pod
Similar to the idea of the proposed solution and option (1), but with adding status condition to the connected pods.

Following CUDN CR localnet spec change, the CUDN controller should add a pod condition on the connected pods,
providing indication about OVN-K network configuration state.
The pod condition should be `true` when the pod network-config is up-to-date.
The pod condition should be `false` when the pod network-config is outdated.

Adding condition to pods is a complementary matter for indicating the subject state, alongside the proposed solution.
It can be done as a follow-up upon necessity and feedback from users.

> #### 3. CUDN CRD status condition

Reviewer 2: I think this is also required IMO. @jcaamano fyi. If a user updates the CRD and it causes either some incompatibility such as: excluding an IP that is already reserved, or some other failure that is caused by the update. Then the CUDN should go into error state. 

If the update to the CRD is valid, but for some reason a node's ovnkube-controller detects some error configuring it, then it should post a substatus failure. The overall CUDN CRD then should also go into error state. I think @npinaeva already has groundwork/plans to do this?

Reviewer 7: Do you mean the NetworkReady condition we were going to report from every node? 
We could do that, but it requires some work for per-node condition reporting + centralized status, and I think currently ovn-k is totally fine with leaving excludedSubnet IP on the pod if it is already assigned (: so some extra changes to report these errors may be needed

Author: I think the first part is covered today, if the CRD spec is invalid or has some bad input, the CRD status will communicate the issue through conditions.
If there was an error with rendering/creating the NAD, will `NetworkCreated` condition will be false.
If there was an error related to the node allocator, `NetworkAllocationSucceeded` condition will be false.

status condition should be true or false.

Adding status condition is a complementary matter for indicating the subject state, alongside the proposed solution.
It can be done as a follow-up upon necessity and feedback from users.

#### 4. Provide collective data for alerting framework
Alert is a notification that is raised based on various data sources in the cluster 
such as object status, event, logs, metrics, etc..

Usually alerting framework enable setting a notification on the platform GUI with a nice message referencing 
to instructions on how to address it.

The proposed pod annotation can serve as the baseline for providing collective data for alerting frameworks,
as the indication for raising an alert, e.g.: when there is at least on pod with outdated network-config.

Alerts are a complementary matter for indicating the subject state, alongside the proposed solution.
It can be done as a follow-up upon necessity and feedback from users.

#### 5. Add support for OVN-K CNI plugin to configure running pods  
As of today OVN-K configure the pod network-stack when a pod is created or deleted.

> Introduce OVN-K CNI plugin using a think plugin architecture, similar to Multus CNI think-plugin

Reviewer 2: you mean "thick" plugin?

we already have something similar with our CNI server. We can detect changes to the NAD there. Alternatively we could implement cni cmd CHECK, that would allow us to reconfigure something in the pod. Anyway, this is something that we can tackle in the future.

Author: Yeah, sorry for this occurred typo.. it was late :see_no_evil: 

> we already have something similar with our CNI server. We can detect changes to the NAD there. Alternatively we could implement cni cmd CHECK, that would allow us to reconfigure something in the pod. Anyway, this is something that we can tackle in the future.

That sound really useful, in particular for MTU changes, which have even greater impact on VM (as we strive for minimal downtime).

Having said that mutating the suggested settings should not break the workloads networking, and they should be able to communicate
with each other.
After the network setting changes, restarting the workloads is needed in order for them to consume the latest network-config.
Once all connected workloads configured with the latest network-config the state should be reconciled and no network-disruption is expected.

Please refer to [Risks Known Limitations and Mitigations](#risks-known-limitations-and-mitigations) section for more details.  

When a CUDN CR localnet spec changes, and the change require updating the corresponding NAD spec.config, the controller should 
add the proposed annotation `k8s.ovn.io/network-out-of-sync` to all connected pods in the selected-namespaces.
In case the annotation is already exists and has items (networks), the controller should append the subject network to the list.

On NAD update the controller should record the update timestamp, for each connected pod and compare the pod creation-timestamp and the NAD update timestamp:
If the pod was created before NAD was updated - add the annotation to the pod.

> When pod created after NAD update, no need to annotate the pod because it consumed the latest network-config.

Reviewer 6: Continuing the previous [thread](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5087#discussion_r2004993490), this is still raceful.

Even if we assume the time known to the UDN controller is in total sync with the on of the api-server, I cannot see how you can assure no pod will enter in-between the steps of NAD update, timestamp saving and processing the pod list. The whole system is async based, nothing is atomic and all actual states are in constant potential transition. The only assured thing that exist is that at some future, we may be able to reach a desired state, and I do not see how this mechanism allows something like this.

Author: We decided to drop spec mutations and buy that the need to indicate users upon network-config change.
The proposal now focus on CRD extensions for localnet topology only.

- As a user I want to be able to connect my workloads (pod/VMs) to the localnet the admin created in my namespace.
- As a user I want my workloads to be able to communicate with each other over the localnet network.
- As a user I want my connected VMs to the localnet network to be able to migrate from one node to another, having its localnet network interface IP address unchanged.
- As an admin I want to be able to perform day2 changes to the localnet topology spec following env changes.
  - Change VLAN, e.g: following the being used has been decommissioned.
  - Changing MTU, e.g: following HW has been changed requiring MTU re-alignment.
  - Change the localnet topology bridge-mapping, e.g.: following a requirement to connect to different interfaces on the node.

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Author: As of today, in case of a NAD is changed workloads should be restart (re-created) in order to consume the new configuration. The same should be done with CUDN CR localnet spec.

Any change that involve the CNI will require a restart in order to trigger the CNI.
 
In case excludeSubnet are changed, restarting the workloads will reconcile the state, and the new excluded IPs will not be assigned to pods.

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Reviewer 7: So I think for MTU, VLAN and excludeSubnets old pods will just keep using old configuration until all pods are re-created, but what is the effect of the physicalNetwork change?

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Reviewer 3: To clarify: it's not until all pods are re-created - each pod will consume the new configuration at creation time. I.e. each workload owner gets to choose if / when to consume the updated configuration.

Does this make sense to you @npinaeva ?

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Reviewer 3: it'll use a different bridge mapping, which might mean it'll use a different ovs bridge.

I don't remember if it'll use a different patch port or not ... We'll need to check. 

Still, thinking this through, I don't think this'll require a workload restart ... It's a localnet port setting ... I guess we'll need to check when does this take effect - i.e. when the network is updated, or when *a* pod is updated. I assume it's when the network is updated.

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Author: I checked how `VLAN` and `physicalNetworkName` mutations take effect and its like you thought:
changing `physicalNetworkName` in the CR takes affect when the NAD changes, no need to restart pods.
And it use different patch-port (the patch-port peer remove from br1 and create on br2, according to the mapping)

One thing was wired though, when `physicalNetworkName` was changed and VLAN remain as is:
Traffic keep going through the original localnet instead of the one we just picked.
The patch-port to br-int remain on the original bridge (ref by the original localnet) instead of being removed and re-created on the desired bridge (ref by the newly specified physicalNetworkName).
In this scenerio, recreating the NAD tirggers recreating the br-int patch-port, result in traffic goes through the desired bridge.

---
See bellow description for more details:

The setup consist of  two additional NICs on each node with corresponding ovs-bridge and localnet bridge-mappings:
```
eth1 (iface) -> ovsbr1 (ovs-bridge) -> localnet1 (bridge-mapping)
eth2 (iface) -> ovsbr2 (ovs-bridge) -> localnet2 (bridge-mapping)
```
1. Mutating VLANs
Create CUDN "test" with: `physicalNetworkName=localnet1`, `vlan=100`.
Created two pods on different nodes, one ping the other.
Change CUDN spec VLAN to 500.
Result: pods can communicate over the new VLAN
- Tcpdump prob on `eth1` shows packets tagged with VID 500.
- When the ovs bridge bridge-ports are configured with VLAN packets are not tagged. There was no traffic disruption.

2. Mutating `physicalNetworkName` & VLAN
Create CUDN "test" with `physicalNetworkName=localnet1`, `vlan=100`
Created two pods on different nodes, one ping the other.
Change the CUDN spec `physicalNetworkName` to `localnet2` and `vlan` to `200`
Result: pods communicate over `localnet2` (ovsbr2 -> eth2) :
- Tcpdump prob on `eth1` (ref by `localnet1`) no traffic.
- Tcpdump prob on `eth2` (ref by `localnet2`) ICMP traffic.
- OVS state shows the patch-port to `br-int`  is removed from `ovsbr1` (ref, by `localnet1`) and exist on `ovsbr2` (ref. by `localnet2`).
- No pod re-creation are needed.
- I got the same result when the ovs bridge bridge-ports were configured with VLAN, traffic goes through the right interface.

3. Mutating `physicalNetworkName` only
Create CUDN "test22" with `physicalNetworkName=localnet1` **(no-vlan)**
Created two pods on different nodes, one ping the other.
Change the CUDN spec `physicalNetworkName` to `localnet2`.
Result: pods communicate over the wrong localnet - **BAD**:
- Tcpdump prob on eth1 (ref by `localnet1`) show traffic. 
- Tcpdump prob on eth2 (ref by `localnet2`) ICMP traffic. 
- OVS state shows the patch port to `br-int` exist on `ovsbr1` (ref by `localnet1`), instead of `ovsbr2` (ref by `localnet2`).
* The patch-port name is generated according to the `netConf` name `cluster_udn_test22` and not the `physicalNetworkName`  (patch-cluster_udn_test22_ovn_localnet-port-to-br-int).

>   - When I use OVN-K IP assignment, I am able to add/remove excluded IP addresses from the subnet the platform uses, 

Reviewer 3: > And it use different patch-port (the patch-port peer remove from br1 and create on br2, according to the mapping)

Yes, of course **if** the bridge changes. 

I wondered what happens if you just change the mapping, but still attach to the same bridge. IIUC, you'll still get a new patch port, since the name of the network has changed. Right ? 

Regarding the network controller not reacting to `physicalNetworkName` changes, I think we have a bug. Please check if this PR works for you: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5125

If so, I'll figure out a way to unit test it, and we can focus on getting it merged.

- `MTU`:
    - Minimum 576 (minimal for IPv4). Maximum: 65536.
    - When Subnets consist of IPv6 CIDR, minimum MTU should be 1280.
- `VLAN`:

>   - `ID`:

Author: > Regarding naming, maybe we'd want to be explicit this would configure an access port by naming the attribute access-port-id . Not a must.

Consider the following alternative for avoid having the mode name in the fields names:
Introduce `mode` field, accepting enum `Access`. 
```
localnet:
  vlan:
    mode: Access
    id: 100
```
For trunk support, extend mode to accepts "Trunk", and add fields for the IDs, for example:
```
localnet:
  vlan:
    mode: Trunk
    trunkIDs: [100, 200,...]
```

With this example, its better to add Mode field now, otherwise the API should do assumptions (missing Mode mean access mode) making the API less explicit and less predictable.
WDYT?
